emulator/src/kreator/preprocessor.rs

```diff
@@ -0,0 +1,532 @@
+        for (key, value) in &labels {
+            owned_line = owned_line.replace(key, &value.to_string());
+        }
+
+        // determine if a variable is being declared by EQU
+        if owned_line.contains("EQU") {
+            let split_name = owned_line.split_once(" ").unwrap();
+            if equate_assignments.contains_key(split_name.0) {
+                return Err("Can't assign a variable more than once using EQU!");
+            }
+            let split_expr = split_name.1.split_once(" ").unwrap();
+            equate_assignments.insert(split_name.0.to_string(), eval_str(split_expr.1.to_string()));
+            owned_line.clear();
+        }
+
+        // determine if a variable is being declared by SET
+        if owned_line.contains("SET") {
+            let split_name = owned_line.split_once(" ").unwrap();
+            let split_expr = split_name.1.split_once(" ").unwrap();
+            set_assignments.insert(split_name.0.to_string(), eval_str(split_expr.1.to_string()));
+            owned_line.clear()
+        }
+
+        // replace values of variables declared by EQU
+        for (key, value) in &equate_assignments {
+            owned_line = owned_line.replace(&format!(" {}", key), &format!(" {}", value));
+        }
+
+        // replace values of variables declared by SET
+        for (key, value) in &set_assignments {
+            owned_line = owned_line.replace(&format!(" {}", key), &format!(" {}", value));
+        }
+
+        // check if conditional is exited (before check for entering since "IF" is contained in "ENDIF")
+        if owned_line.contains("ENDIF") {
+            if !in_conditional {
+                return Err("Every ENDIF must have a corresponding IF");
+            }
+            condition = false;
+            in_conditional = false;
+            owned_line.clear();
+        }
+        // check if conditional is being entered
+        else if owned_line.contains("IF") {
+            in_conditional = true;
+            let condition_str = owned_line.split_once(" ").unwrap().1.to_string();
+            condition = eval_str(condition_str) != 0;
+            owned_line.clear();
+        }
+
+        // check if conditional holds true
+        if in_conditional {
+            if !condition {
+                owned_line.clear();
+            }
+        }
+
+        pc += 1;
+        if !owned_line.is_empty() {
+            preprocessed_code.push(String::from(owned_line.trim()));
+        } else {
+            pc -= 1;
+        }
+    }
+    if in_conditional {
+        return Err("Every IF must be closed");
+    }
+    Ok(preprocessed_code)
+}
+
+fn replace_macros(code: &Vec<String>) -> Result<Vec<String>, &'static str> {
+    let macro_wrap = get_macros(code);
+    if macro_wrap.is_err() {
+        return Err(macro_wrap.unwrap_err());
+    }
+    let macros = macro_wrap.unwrap();
+    let mut macroless_code: Vec<String> = Vec::new();
+    let mut in_macro_declaration = false;
+
+    for line in code {
```

Comment:
This `owned_line.clear()` mechanic seems a bit odd... imo it would be better to just add the line to macroless_code and call `continue` to get to the next iteration, instead of clearing the string. that would reduce the amount of if statements called in the cases where any of these early exit conditions apply and not change anything otherwise. Might be missing something though.